### PR TITLE
Refactored Touch Screen Input Source to Device Manager

### PR DESCRIPTION
**Pull request by StephenHodgson**
Overview
---
- Refactors the Touch Screen input source to a device manager.
- Added Touch Screen Input Device Manager to Mixed Reality Manager.
- Added generic vector2 events for OnGesutureUpdated.
- Fixed an issue with OnGestureCompleted not being raised properly.
- Added Input Source Position Updated with current touch position.
- Added ability to add missing controller types in the Controller Config Profile.
- Improved cursor visibility logic.

Changes
---
- Related to #2680

### CODE REVIEW

**Comment by StephenHodgson** [comment]
Not that I'm aware of.

**Comment by StephenHodgson** [comment]
I forgotten some files

File: Assets/MixedRealityToolkit/InputSystem/Sources/TouchscreenInputSource.cs [hunk 1]
```diff
@@ -1,251 +1,250 @@
-﻿// Copyright (c) Microsoft Corporation. All rights reserved.
-// Licensed under the MIT License. See LICENSE in the project root for license information.
-
-using Microsoft.MixedReality.Toolkit.InputSystem.Pointers;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputSystem/Sources/TouchscreenInputSource.cs]: I should completely remove the this file

File: Assets/MixedRealityToolkit/_Core/Devices/UnityInput/UnityTouchDeviceManager.cs [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+            /// <param name="pointerName"></param>
+            public TouchPointer(string pointerName)
+            {
+                InputSystem = MixedRealityManager.Instance.GetManager<IMixedRealityInputSystem>();
+                PointerId = InputSystem.FocusProvider.GenerateNewPointerId();
+                PointerName = pointerName;
+            }
+
+            /// <inheritdoc />
+            public IMixedRealityInputSystem InputSystem { get; }
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Devices/UnityInput/UnityTouchDeviceManager.cs]: The BaseDeviceManager already provides this, so it's a little redundant here

**Comment by StephenHodgson** [comment]
That's not possible. The unity device manager only updates controllers when the joystick array is updated. Touch screen and mouse input are not part of that array.

File: Assets/MixedRealityToolkit/_Core/Devices/UnityInput/UnityTouchDeviceManager.cs [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+            if (!ActiveTouches.TryGetValue(touch.fingerId, out controller))
+            {
+                var touchPointer = new TouchPointer($"Touch Pointer {touch.fingerId}");
+                var inputSource = InputSystem?.RequestNewGenericInputSource($"Touch {touch.fingerId}", new IMixedRealityPointer[] { touchPointer });
+                touchPointer.InputSourceParent = inputSource;
+
+                controller = new UnityTouchController(TrackingState.NotApplicable, Handedness.Any, inputSource);
+                controller.SetupConfiguration(typeof(UnityTouchController));
+                ActiveTouches.Add(touch.fingerId, controller);
+                Debug.LogWarning("Created new touch controller");
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Devices/UnityInput/UnityTouchDeviceManager.cs]: LogWarnings in active code?

File: Assets/MixedRealityToolkit/_Core/Devices/UnityInput/UnityTouchDeviceManager.cs (hunk 2 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Devices/UnityInput/UnityTouchDeviceManager.cs]: It's part of the pointer interface, so can't really change that.

File: Assets/MixedRealityToolkit/_Core/Managers/MixedRealityManager.cs [hunk 4]
```diff
@@ -943,13 +943,15 @@ private void GetComponentByTypeAndName(Type type, string managerName, out IMixed
 
         private void AddManagersForTheCurrentPlatform()
         {
+            AddManager(typeof(IMixedRealityDeviceManager), new UnityDeviceManager("Unity Device Manager", 10));
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Managers/MixedRealityManager.cs]: I should revert these changes.

File: Assets/MixedRealityToolkit/_Core/Devices/UnityInput/UnityTouchDeviceManager.cs [hunk 5]
```diff
@@ -259,9 +259,14 @@ private void AddTouchController(Touch touch, Ray ray)
...
             {
-                var touchPointer = new TouchPointer($"Touch Pointer {touch.fingerId}");
-                var inputSource = InputSystem?.RequestNewGenericInputSource($"Touch {touch.fingerId}", new IMixedRealityPointer[] { touchPointer });
-                touchPointer.InputSourceParent = inputSource;
+                IMixedRealityInputSource inputSource = null;
+
+                if (InputSystem != null)
+                {
+                    var touchPointer = new TouchPointer($"Touch Pointer {touch.fingerId}");
+                    inputSource = InputSystem?.RequestNewGenericInputSource($"Touch {touch.fingerId}", new IMixedRealityPointer[] { touchPointer });
```
Comment (Jarodshow) [inline Assets/MixedRealityToolkit/_Core/Devices/UnityInput/UnityTouchDeviceManager.cs]: redundant `InputSystem?`

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Devices/UnityInput/UnityTouchDeviceManager.cs]: Great catch

**Comment by davidkline-ms** [comment]
@StephenHodgson, what is blocking this PR?

**Comment by StephenHodgson** [comment]
#2695

File: Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputSystemProfile.cs [hunk 6]
```diff
@@ -53,7 +53,6 @@ public MixedRealityPointerProfile PointerProfile
             get { return pointerProfile; }
             private set { pointerProfile = value; }
         }
-
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputSystemProfile.cs]: need to revert this

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs [hunk 7]
```diff
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.﻿
 
+using System;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs]: nit sort

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputSystemProfileInspector.cs [hunk 8]
```diff
@@ -50,11 +46,22 @@ public override void OnInspectorGUI()
                 return;
             }
 
-            if (GUILayout.Button("Back to Configuration Profile"))
+            if (!MixedRealityManager.Instance.ActiveProfile.IsInputSystemEnabled)
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputSystemProfileInspector.cs]: need to revert this as well

File: Assets/MixedRealityToolkit-Tests/MixedRealityManagerTests.cs [hunk 9]
```diff
@@ -54,7 +54,7 @@ public void Test03_CreateMixedRealityManager()
 
             // Create Test Configuration
             Assert.AreEqual(0, MixedRealityManager.Instance.ActiveProfile.ActiveManagers.Count);
-            Assert.AreEqual(1, MixedRealityManager.Instance.MixedRealityComponents.Count);
+            Assert.AreEqual(3, MixedRealityManager.Instance.MixedRealityComponents.Count);
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit-Tests/MixedRealityManagerTests.cs]: I'll do this in a separate PR

File: Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputSystemProfile.cs (hunk 6 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputSystemProfile.cs]: then get it sorted @StephenHodgson :P

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputSystemProfile.cs]: It's good meow

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityControllerMappingProfile.cs [hunk 10]
```diff
@@ -80,8 +80,9 @@ public GameObject GlobalRightHandModel
             new MixedRealityControllerMapping(8, "Oculus Touch Controller Left", typeof(OculusTouchController), Handedness.Left),
             new MixedRealityControllerMapping(9, "Oculus Touch Controller Right", typeof(OculusTouchController), Handedness.Right),
             new MixedRealityControllerMapping(10, "Oculus Remote Controller", typeof(OculusRemoteController)),
-            new MixedRealityControllerMapping(11, "Generic OpenVR Controller Left", typeof(GenericOpenVRController), Handedness.Left, true),
-            new MixedRealityControllerMapping(12, "Generic OpenVR Controller Right", typeof(GenericOpenVRController), Handedness.Right, true),
+            new MixedRealityControllerMapping(11, "Touch Screen Input", typeof(UnityTouchController), Handedness.Any),
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityControllerMappingProfile.cs]: in the future, we should try to avoid reordering collections in profiles so that we don't create unwanted side effects

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityControllerMappingProfile.cs]: Agreed, but there will be one more change with the addition to the mouse that is in the next PR. I want to look into making it easy to rearrange this in the future.
